### ISSUE TITLE
Clean Install Fails

### ISSUE THREAD

**Issue by skellycode** [paxswill/iocage-zigbee2mqtt#5]
An attempt at a clean installation of this plugin fails.

After some googling around, there seem to be a few different possible causes (not having python 2, incorrect npm version).  I'm not familiar with npm so am not having much luck solving this myself - however if it's a dependancy issue then I think this jail script would require a fix (rather than Z2M).

Happy to help troubleshoot where I can :)

Attempted install console output below:

`root@freenas[~]# iocage fetch -P zigbee2mqtt -g https://github.com/paxswill/iocage-zigbee2mqtt --branch 12.1-RELEASE --name Z2M
Cloning git repository

Branch 12.1-RELEASE does not exist at https://github.com/paxswill/iocage-zigbee2mqtt!
Using "master" branch for plugin, this may not work with your RELEASE
Plugin: zigbee2mqtt
  Official Plugin: false
  Using RELEASE: 12.1-RELEASE
  Using Branch: 12.1-RELEASE
  Post-install Artifact: https://github.com/paxswill/iocage-zigbee2mqtt.git
  These pkgs will be installed:
    - ca_root_nss
    - curl
    - git
    - node
    - npm
Default mac_prefix specified in defaults.json '9c5c8e' is invalid. Using '9e5c8e' mac prefix instead.

Testing Host DNS response to pkg.FreeBSD.org
Testing Z2M's SRV response to pkg.FreeBSD.org
Testing Z2M's DNSSEC response to pkg.FreeBSD.org

Installing plugin packages:
  - ca_root_nss...
  - curl...
  - git...
  - node...
  - npm...

Fetching artifact...
Cloning git repository

Branch 12.1-RELEASE does not exist at https://github.com/paxswill/iocage-zigbee2mqtt.git!
Using "master" branch for plugin, this may not work with your RELEASE

Running post_install.sh
Cloning into '/opt/zigbee2mqtt'...

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
Cloning into '/opt/zigbee2mqtt'...

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
prebuild-install WARN install No prebuilt binaries found (target=16.2.0 runtime=node arch=x64 libc= platform=freebsd)

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gmake: Entering directory '/opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/build'
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/serialport.o
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/serialport_unix.o
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/poller.o
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  SOLINK_MODULE(target) Release/obj.target/bindings.node
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  COPY Release/bindings.node
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gmake: Leaving directory '/opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/build'
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
(node:30843) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
prebuild-install WARN install No prebuilt binaries found (target=16.2.0 runtime=node arch=x64 libc= platform=freebsd)

> @serialport/bindings@9.2.0 install /opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gmake: Entering directory '/opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/build'
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/serialport.o
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/serialport_unix.o
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/bindings/src/poller.o
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  SOLINK_MODULE(target) Release/obj.target/bindings.node
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  COPY Release/bindings.node
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gmake: Leaving directory '/opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/build'
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)

> unix-dgram@2.0.3 install /opt/zigbee2mqtt/node_modules/unix-dgram
> node-gyp rebuild
(node:30954) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)

> unix-dgram@2.0.3 install /opt/zigbee2mqtt/node_modules/unix-dgram
> node-gyp rebuild
(node:31236) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
gmake: Entering directory '/opt/zigbee2mqtt/node_modules/unix-dgram/build'
(node:31236) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/unix_dgram/src/unix_dgram.o
(node:31236) [DEP0150] DeprecationWarning: Setting process.config is deprecated. In the future the property will be read-only.
(Use `node --trace-deprecation ...` to show where the warning was created)
  CXX(target) Release/obj.target/unix_dgram/src/unix_dgram.o
../src/unix_dgram.cc:53:13: warning: unused function 'SetNonBlock' [-Wunused-function]
inline void SetNonBlock(int fd) {
            ^
../src/unix_dgram.cc:65:13: warning: unused function 'SetCloExec' [-Wunused-function]
inline void SetCloExec(int fd) {
            ^
  CXX(target) Release/obj.target/unix_dgram/src/unix_dgram.o
2 warnings generated.
  SOLINK_MODULE(target) Release/obj.target/unix_dgram.node
2 warnings generated.
  COPY Release/unix_dgram.node
2 warnings generated.
gmake: Leaving directory '/opt/zigbee2mqtt/node_modules/unix-dgram/build'
2 warnings generated.
added 309 packages in 12.217s
2 warnings generated.

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
2 warnings generated.

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
sh: tsc: not found

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
npm ERR! code ELIFECYCLE

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
npm ERR! zigbee2mqtt@1.21.0 build: `tsc && node index.js writehash`
npm ERR! spawn ENOENT
npm ERR!

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash
npm ERR! Failed at the zigbee2mqtt@1.21.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

> zigbee2mqtt@1.21.0 build /opt/zigbee2mqtt
> tsc && node index.js writehash

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-08-24T04_10_31_828Z-debug.log
Last 10 lines:

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-08-24T04_10_31_828Z-debug.log
Z2M had a failure
Exception: SystemExit Message: 1
Partial plugin destroyed`

**Comment by paxswill** [comment]
Without digging too deeply into the problem yet, I'm guessing it's related to Z2M's Typescript migration. There was a similar issue on the Z2M project (https://github.com/Koenkk/zigbee2mqtt/issues/8222), and there's been a fix (https://github.com/Koenkk/zigbee2mqtt/commit/62d763614a428f4dbb87ef575c0ede527cbbe647) on the dev branch but it hasn't made it into a release yet. The workaround for upgrading existing jails has been to switch to the dev branch (#4), but I'm not sure how to address that for clean installs. Defaulting everybody to the dev branch doesn't feel like the correct approach, and AFAIK there isn't a way to provide settings from iocage to the plugin (my experience iocage properties [hasn't been great](https://github.com/paxswill/iocage-zigbee2mqtt/commit/f5c57ada93375249340744d82f4df28e9dfb943b) so far).

A possible solution for the next week or so (there should be a Z2M release on Sept 1) would be to change installs that checkout 1.21.0 to instead track dev, with a large notice that they're on dev, give links to this issue and #4, and recommend switching back to the release branch after September. Does that sound alright?

**Comment by skellycode** [comment]
Thank you for the detailed reply!

If you think it's worth investing the time/effort to switch to the dev track just for a few days then I'm happy to test your implementation, however if there's a new release coming next week then it may not be worth it.  I guess it depends how many others are in my boat who excitedly got their coordinator in the mail and then hit a brick wall! 😆

Here are my attempts at resolving the issue thus far (I have forked the repo, and made a few tweaks to allow iocage jail installation to complete, however there are still a bunch of manual steps I've had to take post-install to get it roughly work).

- I tried switching to the dev branch in the code, and pre-installing typescript and removing `npm ci --production` references in `post_install.sh` before attempting to install the jail.  It meant that the jail installation could complete successfully, however there are still issues with startup and the service does not become active
- I then created my own (incredibly barebones) rc.d which allowed Z2M to start - however this is no solution, simply a workaround to address the same issue in the current rc.d script which had references to `npm ci --production` - the only way I could possibly get it to start was to use `npm start`.  Any variation on `run/build/ci/etc` would crash

I hope that's helpful! 😄
